fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AuthorizerStmtVisitor.java

```diff
@@ -227,27 +236,101 @@ public void check(StatementBase statement, ConnectContext context) {
+                            statement.getTableName(), PrivilegeType.INSERT);
+                    insertPrivLevel = PrivilegeLevel.TABLE;
+                } catch (AccessDeniedException e) {
+                    try {
+                        columnNames = statement.getTargetColumnNames() == null ? Collections.singleton("*") :
+                                new HashSet<>(statement.getTargetColumnNames());
+                        Authorizer.checkColumnsAction(context.getCurrentUserIdentity(), context.getCurrentRoleIds(),
+                                statement.getTableName(), columnNames, PrivilegeType.INSERT);
+                        insertPrivLevel = PrivilegeLevel.COLUMN;
+                    } catch (AccessDeniedException exception) {
+                        AccessDeniedException.reportAccessDenied(
+                                statement.getTableName().getCatalog(),
+                                context.getCurrentUserIdentity(), context.getCurrentRoleIds(),
+                                PrivilegeType.INSERT.name(), ObjectType.COLUMN.name(),
+                                String.join(".", statement.getTableName().getTbl(),
+                                        String.join(",", columnNames)));
+                    }
+                }
+            }
+
+            Map<TableName, Relation> allTablesRelations = AnalyzerUtils.collectAllTableAndViewRelations(statement);
+            Map<TableName, Set<String>> allTouchedColumns = AnalyzerUtils.collectAllSelectTableColumns(statement);
             try {
-                Authorizer.checkTableAction(context.getCurrentUserIdentity(), context.getCurrentRoleIds(),
-                        statement.getTableName(), PrivilegeType.INSERT);
-            } catch (AccessDeniedException e) {
-                AccessDeniedException.reportAccessDenied(statement.getTableName().getCatalog(),
-                        context.getCurrentUserIdentity(), context.getCurrentRoleIds(),
-                        PrivilegeType.INSERT.name(), ObjectType.TABLE.name(), statement.getTableName().getTbl());
+                if (insertPrivLevel == PrivilegeLevel.TABLE) {
+                    // no need to check TABLE/COLUMN SELECT privilege when user already has this TABLE's INSERT privilege
+                    allTablesRelations.remove(statement.getTableName());
+                    allTouchedColumns.remove(statement.getTableName());
+                }
+                checkSelectTableAction(context, allTablesRelations);
+            } catch (ErrorReportException e) {
+                if (insertPrivLevel == PrivilegeLevel.COLUMN) {
+                    if (allTouchedColumns.containsKey(statement.getTableName())) {
+                        // no need to check COLUMN SELECT privilege when user already has COLUMN INSERT privilege
+                        Set<String> usedCols = allTouchedColumns.get(statement.getTableName());
```

Comment:
can `usedCols` contain `*`?

Reply:
It is indeed possible, I forgot to consider this scenario. It seems that I need to refine this part of the code.

Reply:
👍🏻